#!/bin/bash
### This script aligns fastq files with bowtie2 and produces a text file with unmapped reads, a text file with depth information, ggplot of the depth and a csv file with low depth positions.
### This script requires an indexed reference genome and the Rscript plot_depth.R.

# Create a variable to store the reference path. Requires a previously created reference with bowtie2-build.
reference="/home/josemari/Desktop/Jose/Reference_sequences/MN908947"
# Create a variable to store the path for Rscripts.
Rscripts="/home/josemari/Desktop/Jose/General_scripts/R"
# Get the name of the fastq files and the episeq consensus.
rawfastq_tag=("CS230192_S14" "CS230195_S17" "CS230196_S18" "CS230201_S23" "CS230191_S13" )
Fw=${rawfastq_tag}"_L001_R1_001.fastq.gz"
Rv=${rawfastq_tag}"_L001_R2_001.fastq.gz"
episeq_cns=${rawfastq_tag}".consensus.fasta"

#COMANDS
#Indexing
#bowtie2-build '$param.ref' sarscov
# Loop through the barcodes
for i in ${rawfastq_tag[@]}; do

# Align the input file to the reference with bowtie2 and write out a txt documment with unnaligned reads.
echo Aligning ${i} with bowtie2...
bowtie2 --un ${i}_not_aligned.txt -x ${reference} -1 ${Fw} -2 ${Rv} -S ${i}.sam
# Convert the sam to bam file and remove sam file.
echo Converting sam to bam file...
samtools view -b ${i}.sam > ${i}.bam
rm ${i}.sam
# Sort the sam file and remove unsorted bam.
echo Sorting bam file...
samtools sort ${i}.bam -o ${i}_srtd.bam
rm ${i}.bam
# Index the sorted bam file.
echo Indexing bam file...
samtools index ${i}_srtd.bam
# Convert conensus fasq from bam file.
echo Making fastq consensus sequence...
samtools mpileup -uf ${reference}.fasta ${i}_srtd.bam | bcftools call -c | vcfutils.pl vcf2fq > ${i}_cns.fq
# Making consensus fasta from consensus fasq...
echo Getting consensus fasta... Bases quality lower than 20 set to N.
seqtk seq -aQ64 -q20 -n N ${i}_cns.fq > ${i}_cns.fasta

# Call the Rscript to find nt differences between the generated consensus and the episeq consensus.
echo "Calling find_nt_diff.R" 
Rscript ${Rscripts}/find_nt_diff.R ${i}_cns.fasta ${episeq_cns}

# Get depth information to text file and remove bam files.
echo Getting depth from bam file...
samtools depth -a -H ${i}_srtd.bam -o ${i}_depth.txt

# Call Rscript for plotting.
Rscript $Rscripts/plot_depth_pairedend.R `pwd`/${i}_depth.txt

# Move the plot and csv files generated by the Rscripts.
mkdir ${i}_results
mv Alignment_changes.csv ${i}_Alignment_changes.csv
mv ${i}_not_aligned.txt ${i}_depth.txt ${i}_depth.pdf ${i}_low_depth_positions.csv ${i}_Alignment_changes.csv ${i}_cns.fasta mafft_aligned.fasta ${i}_results
# Remove unnecesary files.
rm ${i}_srtd.bam ${i}_srtd.bam.bai ${i}_cns.fq fasta_to_align.fasta

done
