#!/bin/bash
### This script aligns fastq files with bowtie2 and produces a text file with unmapped reads, a text file with depth information, ggplot of the depth and a csv file with low depth positions.
### This script requires an indexed reference genome and the Rscript plot_depth.R.

# Create a variable to store the reference path. Requires a previously created reference with bowtie2-build.
reference="/home/josemari/Desktop/Jose/Reference_sequences/Herpesvirus/Herpesvirus3/NC_023677.1"
# Create a variable to store the path for Rscripts.
Rscripts="/home/josemari/Desktop/Jose/General_scripts/R"
# Get the name of the fastq files.
rawfastq_tag="A8_Extract_6_S6_L001"
Fw=${rawfastq_tag}"_R1_001.fastq.gz"
Rv=${rawfastq_tag}"_R2_001.fastq.gz"

#COMANDS
#Indexing
#bowtie2-build {reference_ID}.fasta {reference_ID}
# Align the input file to the reference with bowtie2 and write out a txt documment with unnaligned reads.
echo Aligning ${rawfastq_tag} with bowtie2...
bowtie2 --un ${rawfastq_tag}_not_aligned.txt -x ${reference} -1 ${Fw} -2 ${Rv} -S ${rawfastq_tag}.sam
# Convert the sam to bam file and remove sam file.
echo Converting sam to bam file...
samtools view -b ${rawfastq_tag}.sam > ${rawfastq_tag}.bam
rm ${rawfastq_tag}.sam
# Sort the sam file and remove unsorted bam.
echo Sorting bam file...
samtools sort ${rawfastq_tag}.bam -o ${rawfastq_tag}_srtd.bam
rm ${rawfastq_tag}.bam
# Index the sorted bam file.
echo Indexing bam file...
samtools index ${rawfastq_tag}_srtd.bam
# Convert conensus fasq from bam file.
echo Making fastq consensus sequence...
samtools mpileup -uf ${reference}.fasta ${rawfastq_tag}_srtd.bam | bcftools call -c | vcfutils.pl vcf2fq > ${rawfastq_tag}_cns.fq
# Making consensus fasta from consensus fasq...
echo Getting consensus fasta... Bases quality lower than 20 set to N.
seqtk seq -aQ64 -q20 -n N ${rawfastq_tag}_cns.fq > ${rawfastq_tag}_cns.fasta

# Get depth information to text file and remove bam files.
echo Getting depth from bam file...
samtools depth -a -H ${rawfastq_tag}_srtd.bam -o ${rawfastq_tag}_depth.txt

# Call Rscript for plotting.
Rscript $Rscripts/plot_depth_pairedend.R `pwd`/${rawfastq_tag}_depth.txt

# Move the plot and csv files generated by the Rscripts.
mkdir ${rawfastq_tag}_results
mv ${rawfastq_tag}_not_aligned.txt ${rawfastq_tag}_depth.txt ${rawfastq_tag}_depth.pdf ${rawfastq_tag}_low_depth_positions.csv ${rawfastq_tag}_cns.fasta ${rawfastq_tag}_results
# Remove unnecesary files.
rm ${rawfastq_tag}_srtd.bam ${rawfastq_tag}_srtd.bam.bai ${rawfastq_tag}_cns.fq


