### This script runs featureCounts and DEseq for RNAseq analysis ###

#LIBRARIES:
library(Rsubread)
library(DESeq2)
library(ggplot2)
library(reshape2)
library(topGO)
library(dplyr)
library(plyr)

#FUNCTIONS:
#Create a function to make volcano plots.
volcano <- function(dataset, name){
  
  #Plot log2Fold in x and pvalue on x for basic volcano plot.
  with(dataset, plot(log2FoldChange, -log10(pvalue), pch = 20, main = name, xlim = c(-4, 4)))
  #Add colored points: blue if padj < 0.01, red if log2FC > 1 and padj < 0.05)
  with(subset(dataset, padj < 0.01 ), points(log2FoldChange, -log10(pvalue), pch = 20, col = "blue"))
  with(subset(dataset, padj < 0.01 & abs(log2FoldChange) > 2), points(log2FoldChange, -log10(pvalue), pch = 20, col = "red"))
  
  return()
}
#Create a function that extracts the differentially expressed genes.
get_DE_genes <- function(dataset, foldchange_threshold, p_value_threshold){
  
  #Set a filter being logfold absolute values higher than the threshold and p value lower than threshold.
  filter = which(abs(dataset$log2FoldChange) > foldchange_threshold & dataset$padj < p_value_threshold)
  #Take from the dataset only the rows that fit the condition.
  DE_genes = dataset[filter,]
  #Get the indices (positions) of the rows sorted decreasingly using the logfold.
  index = sort.int(DE_genes$log2FoldChange, decreasing = TRUE, index.return = TRUE)$ix
  #Sort the genes according to the index to get the upregulated at the top and downregulated at the end.
  DE_genes = DE_genes[index,]
  
  return(DE_genes)
}

#SETTINGS:
#Set working directory to the file's folder.
setwd("/home/josemari/Desktop/phoPR_analysis/phoPRComplMZC_DEanalysis/X204SC21070199-Z01-F001/raw_data/")
#Set the path for the annotation file.
annotation = "/home/josemari/Desktop/phoPR_analysis/Mbovis_reference_genome/LT708304.gff"

#FEATURECOUNTS.
#Run featurecounts and store the output in a variable.
#filenames <- c("R1_srtd.bam", "R2_srtd.bam", "R3_srtd.bam", "R4_srtd.bam", "R5_srtd.bam", "R6_srtd.bam", "R7_srtd.bam", "R8_srtd.bam", "R9_srtd.bam", "R10_srtd.bam", "R11_srtd.bam", "R12_srtd.bam")
#ftCnts_table = featureCounts(filenames, isPairedEnd=TRUE, annot.ext = annotation, isGTFAnnotationFile = TRUE, GTF.featureType = "CDS", GTF.attrType = "locus_tag", nthreads=3, strandSpecific = 0)
#Create a file to store the counts generated by featureCounts, so we dont need to run it every time we do the analysis. Separation will be tab and we want header and row names.
#write.table(ftCnts_table$counts, file = "ExpressionCountsLocus_tagFormat.txt", sep = "\t",row.names = TRUE, col.names = TRUE)

#LOAD DATA.
#Read the table and store it in a variable.
ExpCounts <- read.table("ExpressionCountsLocus_tagFormat.txt", header = TRUE, sep = "\t", stringsAsFactors = FALSE, row.names = 1)
#The rownames will become only the geneID without the tag "geneID" repeated in all of them using the sapply function to split hte names in two and return only the second part which is the gene ID.
#rownames(ExpCounts) = sapply(rownames(ExpCounts), function(x) unlist(strsplit(x, ":"))[2])
#Read a previously created table with the names of the samples and their tags.
samples <- read.table("ExpressionMetadata.txt", header = TRUE, sep = "\t", stringsAsFactors = FALSE)

#DESEQ:
#Create the Deseq2 object with the counts table and the names table.
dds <- DESeqDataSetFromMatrix(countData = ExpCounts, colData = samples, design = ~ Design)
#Run the DESeq function in the DESeq object.
dds <- DESeq(dds)

#PCA plot.
#Variance stabilizing transformation (normalize data using variance between samples).
vsdata <- vst(dds, blind = FALSE)
#Rename the columns with the Design tags instead of the bam fles that are by default.
colnames(vsdata) <- samples$Design
plotPCA(vsdata, intgroup = "Design")

#COMPARISONS.
#Make results tables specifying the comparison. Put the the "different" as numerator so upregulated genes stay on top.
res_Dpho_WT = results(dds, contrast = c("Design", "Dpho", "WT"))
res_MCZBov_Dpho = results(dds, contrast = c("Design", "MCZBov", "Dpho"))
res_MCZTB_Dpho = results(dds, contrast = c("Design", "MCZTB", "Dpho"))
write.table(res_Dpho_WT, file = "resWTvsDpho.txt", sep = "\t",row.names = TRUE, col.names = TRUE)
#res_MCZBov_MCZTB = results(dds, contrast = c("Design", "MCZBov", "MCZTB"))
#res_MCZBov_Dpho = results(dds, contrast = c("Design", "MCZBov", "Dpho"))
#res_MCZTB_Dpho = results(dds, contrast = c("Design", "MCZTB", "Dpho"))
#Visualization.
#head(res_Dpho_WT, tidy = TRUE)

#Make volcano plots for each comparison.
volcano(res_Dpho_WT, "Dpho vs WT")
volcano(res_MCZBov_WT, "MCZBov vs WT")
volcano(res_MCZTB_WT, "MCZTB vs WT")
#volcano(res_CBov_CTB, "CBov vs CTB")
#volcano(res_CBov_Dpho, "CBov vs Dpho")
#volcano(res_CTB_Dpho, "CTB vs Dpho")

#Get the differentially expressed genes for the desired comparison.
#Make it restrictive (foldchange 3, pvalue 0.01 for getin few genes for heatmap, increase it later for pathway searching).
DE_Dpho_WT = get_DE_genes(res_Dpho_WT, 1.5, 0.05)
write.table(DE_Dpho_WT, file = "DE_WTvsDpho_Locus_tagFormat.txt", sep = "\t",row.names = TRUE, col.names = TRUE)

DE_MCZBov_Dpho = get_DE_genes(res_MCZBov_WT, 3, 0.01)
DE_MCZTB_Dpho = get_DE_genes(res_MCZTB_WT, 3, 0.01)


#Heatmaps.
#Convert the variance stabilized DESeq object into a data frame and get row names in a variable.
vsDataFrame <- as.data.frame(assay(vsdata))
vsDataFrame$GeneID <- rownames(vsDataFrame)
#Get the rows from the DF that are DE in a certain condition and specify the colomns(samples) in which we want to see the genes.
#Important to include last column as we created that variable with the names of the genes.
temp = vsDataFrame[rownames(DE_MCZTB_Dpho), c(4:9,13)]
#Reformat the data frame in an easy way to read for the heatmap.
temp_long <- melt(temp, id.vars = c("GeneID"))
#Make the geneID rows reorganized so the upregulated genes end up on top.
temp_long$GeneID <- factor(temp_long$GeneID, levels = rev(rownames(DE_MCZTB_Dpho)))
#Plot the heatmap.
ggplot(temp_long, aes(x = variable, y = GeneID, fill = value)) + geom_tile() + xlab("Samples") + ylab("Gene ID") + ggtitle("Dpho vs MCZTB DE GENES")+ theme(legend.title = element_text(size = 9), plot.title = element_text(size = 12), axis.title = element_text(size = 10, face = "bold"), axis.text.y = element_text(size = 8), axis.text.x = element_text(angle = 65, hjust = 1)) + labs(fill = "Reads")

#Pathway analysis. Need to clean the gff and results data first.
#Relax the threshold to get more genes for pathway analysis.
path_data <- get_DE_genes(res_Dpho_Control, 1.5, 0.1)
#Remove unnecessesary columns.
path_data$baseMean <- NULL
path_data$lfcSE <- NULL
path_data$stat <- NULL
path_data$pvalue <- NULL
#Create a file with the gene names, foldchange and pvalue.
write.table(path_data, file = "results_pho.txt", sep = "\t",row.names = TRUE, col.names = TRUE)
#Change the gene ID by the gene name.
#Move to the location of the gff.
setwd("/home/josemari/phoPR_analysis/Mbovis_reference_genome/")
#Read the gff and create a translation matrix with the IDs, the genesIDs and names only.
gff.data = read.table("LT708304.gff", sep = "\t", head = FALSE, stringsAsFactor = FALSE);
translation.table = matrix(ncol = 3, nrow = 0);
colnames(translation.table) <- c("GeneID", "ID", "gene");
for(i in 1:nrow(gff.data)){
  if(gff.data[i,3] == "gene"){
    entries = unlist(strsplit(gff.data[i,9], ";"));
    geneID = unlist(strsplit(entries[1], "="))[2];
    ID = unlist(strsplit(geneID, ":"))[2];
    gene = unlist(strsplit(entries[2], "="))[2];
    translation.table = rbind(translation.table, c(geneID, ID, gene));
  }
}
#Go back to the working folder.
setwd("/home/josemari/phoPR_analysis/phoPR_DEanalysis/raw_data/")
#Read the results of the pho vs control.
results_pho = read.table("results_pho.txt", head = TRUE, stringsAsFactor = FALSE, sep = "\t");
#Substitute the IDs for the gene names using the translation matrix.
gene.names = c();
for(name in rownames(results_pho)){
  index = which(translation.table[,"ID"] == name);
  gene.names = c(gene.names, translation.table[index, "gene"]);
}
rownames(results_pho) <- gene.names;
results_pho = cbind(results_pho, rownames(results_pho));
colnames(results_pho) <- c("log2FoldChange", "padj", "GeneID")

#Making gene mapping from AmiGO 2 output: http://amigo.geneontology.org/amigo/search/annotation.
setwd("/home/josemari/phoPR_analysis/Mbovis_reference_genome/")
panther.data = read.table("Mycobacterium.txt", sep = "\t", head = FALSE, stringsAsFactor = FALSE);
setwd("/home/josemari/phoPR_analysis/phoPR_DEanalysis/raw_data/")

GO.list = list()
for(i in 1:nrow(panther.data)){
  if(is.null(GO.list[[panther.data[i, 3]]])){
    GO.list[[panther.data[i, 3]]] = c(panther.data[i, 5])
  }else{
    GO.list[[panther.data[i, 3]]] = c(GO.list[[panther.data[i, 3]]], panther.data[i, 5])
  }
}

con <- file("gene2GO.txt", "w");
for(gene in names(GO.list)){
  write(paste0(gene, "\t", paste(GO.list[[gene]], collapse=","), ","), file=con);
}
close(con)
# Read in some table with gene ids, for example expressions
coreGenes <- results_pho

# Get gene mappings to GO terms
# This table has to look like this:
# FN03409	GO:0009514,GO:0004451,GO:0046872,GO:0006097,GO:0006099,
# FN23295	GO:0009507,GO:0022626,GO:0009514,GO:0005739,
# FN06463	GO:0009514,GO:0004474,GO:0006097,GO:0006099,
# FN16456	GO:0048046,GO:0009507,GO:0009570,GO:0005829,GO:0016020,GO:0005777,
# FN00210	GO:0005576,GO:0003993,GO:0046872,
# FN00543	GO:0005788,GO:0003756,GO:0045454,
# FN26703	GO:0005829,GO:0016020,GO:0005730,GO:0005634,GO:0005524,GO:0004612,
# FN16210	GO:0005829,GO:0016020,GO:0005739,GO:0005634,GO:0005886,GO:0009506,
# FN20048	GO:0009507,GO:0016020,GO:0005777,GO:0009536,GO:0004069,GO:0080130,
# FN26629	GO:0022625,GO:0003723,GO:0003735,GO:0000027,GO:0006412,

geneID2GO <- readMappings("gene2GO.txt")
geneNames <- names(geneID2GO)

# Get the list of genes of interest
myInterestingGenes <- coreGenes$GeneID
geneList <- factor(as.integer(geneNames %in% myInterestingGenes))
names(geneList) <- geneNames
head(geneList)

GOdata <- new("topGOdata", ontology = "BP", allGenes = geneList,
              annot = annFUN.gene2GO, gene2GO = geneID2GO)
# Run topGO with elimination test
resultTopGO.elim <- runTest(GOdata, algorithm = "elim", statistic = "Fisher" )
allRes <- GenTable(GOdata, elimKS = resultTopGO.elim,
                   orderBy = "elimKS", 
                   topNodes = 356)
write.table(allRes, file = "topGO_results.txt", sep = "\t", quote = F, col.names = T, row.names = F)
